### ISSUE TITLE
Insert NONMEM compiler info in lst file

### ISSUE THREAD

**Issue by seth127** [metrumresearchgroup/babylon#198]
Originally posted by @billkmet in rbabylon [here](https://github.com/metrumresearchgroup/rbabylon/issues/271) but this is actually a babylon issue (per @dpastoor comment [here](https://github.com/metrumresearchgroup/rbabylon/issues/271#issuecomment-722424151)).

# Original post

For traceability for regulatory authorities, we would like to capture the gfortran version, and compiler options. The compiler options can be found at the top of the NMFE file used to run NONMEM by rbabylon. We woiuld also like the gfortran version captured as well.  Basically output from gfortran --version.  So the information might look something like below:

GNU Fortran (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
f=gfortran
o=' -w -O3 -ffast-math -ffpe-summary=none '
opstat=' -static '

This can be added to lst file at beginning or end, whichever is more convenient, but each like shouod be preceeded by a ";" so this is treated as a comment in the lat file.  So something like the followoing.

; GNU Fortran (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
; f=gfortran
; o=' -w -O3 -ffast-math -ffpe-summary=none '
; opstat=' -static '

@devinp let me know if more info is needed.

**Comment by kylebaron** [comment]
@seth127  - We're talking about tracking fortran compiler version in the science advisory meeting today. What do you think about going forward with this?

Or consider another proposal that would track gfortran version in one of the yaml files if that would be easier?

Thanks!
Kyle

**Comment by seth127** [comment]
I think including this in the `bbi_config.json` is the right place for it. 

@kyleam do we need more info here about what we would need to capture? I, for one, am not clear about this:

> The compiler options can be found at the top of the NMFE file used to run NONMEM

@kylebaron how much do you know about where/how to extract those compiler options?

**Comment by kyleam** [comment]
> I think including this in the `bbi_config.json` is the right place for it.

I agree conceptually, but based on the [original issue](https://github.com/metrumresearchgroup/bbr/issues/271#issuecomment-722422556), it sounds like that won't do: "The capture of this info is not for our records but info for regulatory authorities. At the end of a project, all we send to the authorities is the data and ctl and lst files for NONMEM."  Has something changed here?

**Comment by seth127** [comment]
Hmm, fair question and I think we'd defer to you @kylebaron (and maybe need to consult further with other folks). That said, I really don't love the idea of modifying the `.lst` file directly, unless absolutely necessary.

**Comment by kyleam** [comment]
> That said, I really don't love the idea of modifying the `.lst` file directly, unless absolutely necessary

Yes, I very much agree.

> @kyleam do we need more info here about what we would need to capture? I, for one, am not clear about this:
> 
> > The compiler options can be found at the top of the NMFE file used to run NONMEM

I'm not sure either, but my understanding at this point:

 * read the nm script we're using, extracting out the information on lines 3-6 below to get the compile options and fortran executable
 * use that to call `{f} --version` to get the fortran version 

```
$ head -n6 /opt/NONMEM/nm75/run/nmfe75 | nl
     1  #!/bin/bash
     2  dir=/opt/NONMEM/nm75
     3  f=gfortran
     4  o=' -w -O3 -ffast-math -ffpe-summary=none '
     5  opstat=' -static '
     6  lib=
```

**Comment by kylebaron** [comment]
Just getting back into this: 

I'd rather just keep this info in one of the bbr files (json or yaml) too. If we need to send this information to regulators, then we can just pack up whatever json or yaml files are needed too. 

I don't have any perspective on capturing compiler options. IMO most of the time you'd just report what compiler was used; I've never seen all the options listed out. 

But if we track the Metworx blueprint number in one of these files as well, could we start tracking gfortran version (guessing we already do) and compiler options for the different NONMEM versions? That way you could still get at them through the blueprint number. 

Thanks.

**Comment by seth127** [comment]
> I'd rather just keep this info in one of the bbr files (json or yaml) too. If we need to send this information to regulators, then we can just pack up whatever json or yaml files are needed too.

Ok cool, I'm glad we feel this way. (I assume you mean "one of the bbi files" but either way, that's an implementation detail)

I would push for capturing this in `bbi_config.json` because it seems in line with the other info in there. Unless there are dissenting opinions on this, that would be my call.

In terms of what exactly we capture, I'm open to anything and everything that's been mentioned. As long as we can reliably grab it and reasonably encode it in `.json`, I don't see why we shouldn't include it. @kyleam we probably want to start a more formal list somewhere (maybe in the issue description, so it's easier to reference? or in another fresh issue?) but this is what I've got so far:

* compiler and compiler version (will this always be gfortran?)
* compiler options (from nmfe file? can we actually grab these reliably?)
* Metworx version, if present
* Any other OS- or hardware-related info that we don't currently grab? 
  * Seems like grabbing the OS and version might be a good idea in general
  * but I'm really thinking about things that could be used to triangulate other system info that we might care about later
